autogen/oai/client.py

```diff
@@ -52,6 +53,246 @@
 LEGACY_CACHE_DIR = ".cache"
 
 
+class ModelClient(Protocol):
+    """
+    A client class must implement the following methods:
+    - create must return a response object that implements the ModelClientResponseProtocol
+    - cost must return the cost of the response
+    - get_usage must return a dict with the following keys:
+        - prompt_tokens
+        - completion_tokens
+        - total_tokens
+        - cost
+        - model
+
+    This class is used to create a client that can be used by OpenAIWrapper.
+    The response returned from create must adhere to the ModelClientResponseProtocol but can be extended however needed.
+    The message_retrieval method must be implemented to return a list of str or a list of messages from the response.
+    """
+
+    RESPONSE_USAGE_KEYS = ["prompt_tokens", "completion_tokens", "total_tokens", "cost", "model"]
+
+    class ModelClientResponseProtocol(Protocol):
+        class Choice(Protocol):
+            class Message(Protocol):
+                content: str | None
+
+        choices: List[Choice]
+        model: str
+
+    def create(self, params) -> ModelClientResponseProtocol:
+        ...  # pragma: no cover
+
+    def message_retrieval(
+        self, response: ModelClientResponseProtocol
+    ) -> Union[List[str], List[ModelClient.ModelClientResponseProtocol.Choice.Message]]:
+        """
+        Retrieve and return a list of strings or a list of Choice.Message from the response.
+
+        NOTE: if a list of Choice.Message is returned, it currently needs to contain the fields of OpenAI's ChatCompletion Message object,
+        since that is expected for function or tool calling in the rest of the codebase at the moment, unless a custom agent is being used.
+        """
+        ...  # pragma: no cover
+
+    def cost(self, response: ModelClientResponseProtocol) -> float:
+        ...  # pragma: no cover
+
+    @staticmethod
+    def get_usage(response: ModelClientResponseProtocol) -> Dict:
```

Comment:
the idea is that it can be left empty if someone doesn't want to track usage

Reply:
In general, it is better to have values optional than fields maybe defined. Once static type checks are added to project, it is easier to handle None values but no type checking can deal with missing attributes. Eventually you would end up with way too many type ignore statements and lesser quality. Pydantic BaseModel is even better because it is tighter and event easier to statically check. Dictionaries are more flexible, but there is no need for flexibility here.

Reply:
Pydantic class with optional types for fields would still be better, it give you more expressive power for type checking and autocomplete.

Reply:
Why not make it return Optional if the intention is for implementing this to be optional? 

I really like this Usage class because that way if you get a value you know exactly what it is